### DISCUSSION TITLE
Storybook 8 - ForwardRef breaks Typescript JSDoc / autodoc description

### DISCUSSION THREAD

**Discussion by tobeyboe** [storybookjs/storybook#26388] [Help]
### Summary

Having issues with using React.forwardRef in Storybook 8, it no longer picks up the JSDoc from the type/interface. 
![image](https://github.com/storybookjs/storybook/assets/18659342/22e9c9b4-9037-4758-b4a8-9f11e460483d)

In Storybook 7.6.17 (and 8 without forwardRef):
![image](https://github.com/storybookjs/storybook/assets/18659342/131c98f8-0846-4bd5-975b-43761f9bcad3)

Recreated using `create-next-app` and Storybook Example/Button (`Button.tsx`)

Storybook 8 Example Button but with forwardRef.
```jsx
/**
 * Primary UI component for user interaction
 */
export const Button = React.forwardRef<HTMLButtonElement, ButtonProps>(
  (
    { primary = false, size = "medium", backgroundColor, label, ...props },
    ref
  ) => {
    const mode = primary
      ? "storybook-button--primary"
      : "storybook-button--secondary";
    return (
      <button
        ref={ref}
        type="button"
        className={["storybook-button", `storybook-button--${size}`, mode].join(
          " "
        )}
        {...props}
      >
        {label}
        <style jsx>{`
          button {
            background-color: ${backgroundColor};
          }
        `}</style>
      </button>
    );
  }
);

Button.displayName = "Button";
```

### Additional information

Storybook 8 Example Button but with forwardRef.
/**
 * Primary UI component for user interaction
 */
export const Button = React.forwardRef<HTMLButtonElement, ButtonProps>(
  (
    { primary = false, size = "medium", backgroundColor, label, ...props },
    ref
  ) => {
    const mode = primary
      ? "storybook-button--primary"
      : "storybook-button--secondary";
    return (
      <button
        ref={ref}
        type="button"
        className={["storybook-button", `storybook-button--${size}`, mode].join(
          " "
        )}
        {...props}
      >
        {label}
        <style jsx>{`
          button {
            background-color: ${backgroundColor};
          }
        `}</style>
      </button>
    );
  }
);

Button.displayName = "Button";

**Comment by hailwood** [comment]
I found this seems to work okay if we do

```tsx
forwardRef((props: MyProps, ref: ForwardedRef<HTMLButtonElement>) => {
  //...
});
```

Instead of 
```tsx
forwardRef<HTMLButtonElement, MyProps>((props, ref) => {
  //...
});
```

**Reply by tobeyboe** [reply]
That might work, but I dont know how I feel about rewriting existing components just to satisfy Storybook. 
That seems like the wrong way around. Especially as it used to work in Storybook 7 and is a valid syntax.  
If I had to change something in the *.stories.tsx that would be fine, as thats within Storybooks domain. If that makes sense.

**Comment by valentinpalkovic** [comment]
In Storybook 8, we are using `react-docgen` as the default parser for component analysis like describe here: https://github.com/storybookjs/storybook/blob/next/MIGRATION.md#react-docgen-component-analysis-by-default

We feel react-docgen is the right tradeoff for most React projects. However, if you need the full fidelity of react-docgen-typescript, you can opt-in using the following setting in .storybook/main.js:

```tsx
// .storybook/main.js|ts
export default {
  typescript: {
    reactDocgen: "react-docgen-typescript",
  },
};
```

Additionally, if this kind of issue isn't reported yet, feel free to report it here: https://github.com/reactjs/react-docgen/issues

**Reply by tobeyboe** [reply]
Ah, thank you very much. That seems to fix it.
That gives "Types of property 'typescript' are incompatible." , but that seems to be an known issue.
https://github.com/storybookjs/storybook/issues/25736

```ts
import type { StorybookConfig } from "@storybook/nextjs";
import path from "path";

const config: StorybookConfig = {
...
  typescript: {
    reactDocgen: "react-docgen-typescript",
  },
};
export default config;
```

**Reply by tobeyboe** [reply]
There seems to be an issue with using `react-docgen-typescript` as that breaks the standard "show code" output.
![image](https://github.com/storybookjs/storybook/assets/18659342/36051aa0-7af5-4641-9dcc-72b964dbf094)

Recreated with standard next.js and storybook 8, standard Button.tsx.
```jsx
const config: StorybookConfig = {
...
  typescript: {
    reactDocgen: "react-docgen-typescript",
  },
};
export default config;
```

```jsx
<(param)=>{
    let { primary = false, size = "medium", backgroundColor, label, ...Props } = param;
    const mode = primary ? "storybook-button--primary" : "storybook-button--secondary";
    return /*#PURE*/ (0,reactJsxDevRuntimeWEBPACKIMPORTEDMODULE0.JsxDEV)("button", {
        type: "button",
        ...Props,
        className: styledJsxStyleWEBPACKIMPORTEDMODULE1Default().Dynamic([
            [
                "7466339eba5534fe",
                [
                    backgroundColor
                ]
            ]
        ]) + " " + (props && props.ClassName != null && props.ClassName || [
            "storybook-button",
            "storybook-button--".Concat(size),
            mode
        ].Join(" ") || ""),
        children: [
            label,
            /*#PURE*/ (0,reactJsxDevRuntimeWEBPACKIMPORTEDMODULE0.JsxDEV)((styledJsxStyleWEBPACKIMPORTEDMODULE1Default()), {
                id: "7466339eba5534fe",
                dynamic: [
                    backgroundColor
                ],
                children: "button.Jsx-style-dynamic-selector{background-color:".Concat(backgroundColor, "}")
            }, void 0, false, void 0, undefined)
        ]
    }, void 0, true, {
        fileName: "/Users/torbjornbo/code/playground/storybook-800Official/src/stories/Button.Tsx",
        lineNumber: 44,
        columnNumber: 5
    }, undefined);
}
  label="Button"
  onClick={() => {}}
  primary
/>
```

So Im kind of stuck between a rock and a hard place. Either lose all type description or all normal "show code"

**Reply by hailwood** [reply]
For other web users that find this, check this related issue https://github.com/storybookjs/storybook/issues/20920#issuecomment-1991262588
